src/MPAGSCipher/PlayfairCipher.cpp

```diff
@@ -0,0 +1,173 @@
+
+    // Make sure the key is uppercase
+    std::transform(std::begin(key_), std::end(key_), std::begin(key_), ::toupper);
+
+    // Remove non-alpha chars
+    auto iter = std::remove_if(std::begin(key_), std::end(key_), [] (char x) {return !std::isalpha(x); });
+
+    key_.erase(iter, std::end(key_));
+
+    // Change J -> I
+    std::transform(std::begin(key_), std::end(key_), std::begin(key_), [] (char x) {return (x == 'J') ? 'I' : x;});
+
+    // Remove duplicate letters
+    std::string foundChar{};
+
+    auto dupeChar = [&] (char x) {if (std::string::npos == foundChar.find(x)) {
+        foundChar+=x;
+        return false;
+    } else {
+        return true;
+    }}; 
+
+    auto iter2 = std::remove_if(std::begin(key_), std::end(key_), dupeChar);
+
+    key_.erase(iter2, std::end(key_));
+
+    // Store the coords of each letter
+    // Store the playfair cipher key map
+    int j = 0;
+    int k = 0;
+    int keySize = key_.size();
+
+    for (int i = 0; i < keySize; i++) {
+        if (j < 5) {
+            std::pair<char, std::pair<int, int>> charCoord{key_[i], {k,j}};
+            std::pair<std::pair<int, int>, char> coordChar{{k,j}, key_[i]};
+            keyMatrix.insert(charCoord);
+            coordMatrix.insert(coordChar);
+            j++;
+        } else {
+            j = 0;
+            k++;
+            std::pair<char, std::pair<int, int>> charCoord{key_[i], {k,j}};
+            std::pair<std::pair<int, int>, char> coordChar{{k,j}, key_[i]};
+            keyMatrix.insert(charCoord);
+            coordMatrix.insert(coordChar);
+            j++;
+        }
+    } 
+}
+
+std::string PlayfairCipher::applyCipher(const std::string& inputText, const CipherMode cipherMode) const {
+
+    std::string cipherText = inputText;
+    int cipherSize = cipherText.size();
+    // Change J → I
+    std::transform(std::begin(cipherText), std::end(cipherText), std::begin(cipherText), [] (char x) {return (x == 'J') ? 'I' : x;});
+
+    // If repeated chars in a digraph add an X or Q if XX
+    for(int i=1; i<cipherSize;) {
+        if (cipherText[i-1] == 'X' && cipherText[i] == 'X') {
+            cipherText.insert(i, 1, 'Q');
+            i+=2;
+        } else if (cipherText[i-1] == cipherText[i]) {
+            cipherText.insert(i, 1, 'X');
+            i+=2;
+        } else {i+=2;}
+    }
+
+    // if the size of input is odd, add a trailing Z
+    if (cipherText.size() % 2 != 0) {
+        cipherText+='Z';
+    }
+
+    switch (cipherMode) {
+        case CipherMode::Encrypt:            
+            // Loop over the input in Digraphs
+            
+            // - Find the coords in the grid for each digraph
+
+            // - Apply the rules to these coords to get 'new' coords
+
+            // - Find the letter associated with the new coords
+            cipherSize = cipherText.size();
+            for(int i=1; i<cipherSize; i+=2) {
+                auto charCoord1 = (*keyMatrix.find(cipherText[i-1])).second;
```

Comment:
You can just use `[]` or (even better) `at` here as then you will be returned the `pair` rather than an iterator. e.g., this becomes:
```
auto charCoord1 = keyMatrix.at(cipherText[i-1]).second
```